python/pyarrow/_cuda.pyx

```diff
@@ -0,0 +1,760 @@
+        return pyarrow_wrap_cudabuffer(
+            shared_ptr[CCudaBuffer](make_shared[CCudaBuffer](parent,
+                                                             offset_, size)))
+
+    def to_pybytes(self):
+        return self.copy_to_host().to_pybytes()
+
+    def __getbuffer__(self, cp.Py_buffer* buffer, int flags):
+        # copy_to_host() is needed otherwise, the buffered device
+        # buffer would contain data pointers of the device
+        raise NotImplementedError('CudaBuffer.__getbuffer__')
+
+    def __getreadbuffer__(self, Py_ssize_t idx, void** p):
+        raise NotImplementedError('CudaBuffer.__getreadbuffer__')
+
+    def __getwritebuffer__(self, Py_ssize_t idx, void** p):
+        raise NotImplementedError('CudaBuffer.__getwritebuffer__')
+
+
+cdef class HostBuffer(Buffer):
+    """Device-accessible CPU memory created using cudaHostAlloc.
+
+    To create a HostBuffer instance, use
+
+      cuda.allocate_host_buffer(<nbytes>)
+
+    The memory is automatically freed when HostBuffer instance is
+    deleted.
+
+    Note: after freeing HostBuffer memory, the instance should be
+    discarded as unusable.
+
+    """
+
+    def __init__(self):
+        raise TypeError("Do not call HostBuffer's constructor directly,"
+                        " use `allocate_host_buffer` function instead.")
+
+    cdef void init_host(self, const shared_ptr[CCudaHostBuffer] &buffer):
+        self.host_buffer = buffer
+        self.init(< shared_ptr[CBuffer] > buffer)
+
+    @property
+    def size(self):
+        return self.host_buffer.get().size()
+
+
+cdef class BufferReader(NativeFile):
+    """File interface for zero-copy read from CUDA buffers.
+
+    Note: Read methods return pointers to device memory. This means
+    you must be careful using this interface with any Arrow code which
+    may expect to be able to do anything other than pointer arithmetic
+    on the returned buffers.
+    """
+    def __cinit__(self, CudaBuffer obj):
+        self.buffer = obj
+        self.reader = new CCudaBufferReader(self.buffer.buffer)
+        self.rd_file.reset(self.reader)
+        self.is_readable = True
+        self.closed = False
+
+    def read_buffer(self, nbytes=None):
+        """Read into new device buffer.
+
+        Parameters
+        ----------
+        nbytes : int, default None
+          Specify the number of bytes to read. Default: None (read all
+          remaining bytes).
+
+        Returns
+        -------
+        cbuf : CudaBuffer
+          New device buffer.
+        """
+        cdef:
+            int64_t c_nbytes
+            int64_t bytes_read = 0
+            shared_ptr[CCudaBuffer] output
+
+        if nbytes is None:
+            c_nbytes = self.size() - self.tell()
+        else:
+            c_nbytes = nbytes
+
+        with nogil:
+            check_status(self.reader.Read(c_nbytes,
+                                          < shared_ptr[CBuffer]* > &output))
```

Comment:
I'm curious whether it's supported to do a cast of a pointer to `shared_ptr` (as opposed to casting the `shared_ptr` itself).

Reply:
From pyarrow.cuda perspective, the Arrow C++ interface (in terms of Cython)
```CStatus Read(int64_t nbytes, shared_ptr[CBuffer]* out)```
should be changed to
```CStatus Read(int64_t nbytes, shared_ptr[CCudaBuffer]* out)```
Then there would no casting needed. Not sure why CBuffer is used, the data pointers are device pointers anyway.